### PR TITLE
Implement protobuf adapter for wire serialization

### LINKED ISSUE #311: Add protobuf wire encoding
We recently introduced an abstraction that makes it possible to inject a custom wire encoding. In particular, this was done for enabling protobuf based encoding. The next step is to write such an encoder.

- [x] #316 
- [x] #317 
- [ ] #318

### PR DESCRIPTION
**Pull request by manoranjith**
Closes #311

### CODE REVIEW

File: client/proposalmsgs_test.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ import (
 	"perun.network/go-perun/channel/test"
 	"perun.network/go-perun/client"
 	clienttest "perun.network/go-perun/client/test"
+	_ "perun.network/go-perun/wire/perunio/serializer" // wire serialzer init
```
Comment (matthiasgeihs) [inline client/proposalmsgs_test.go:27]: Could we move this file to `wire/perunio/test` since it seems to be specific to perunio?

File: wire/controlmsgs_test.go [hunk 2]
```diff
@@ -17,6 +17,7 @@ package wire_test
 import (
 	"testing"
 
+	_ "perun.network/go-perun/wire/perunio/serializer" // wire serialzer init
```
Comment (matthiasgeihs) [inline wire/controlmsgs_test.go:20]: Could we move this file to wire/perunio/test since it seems to be specific to perunio?

File: wire/protobuf/serializer.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	"perun.network/go-perun/wire"
+)
+
+// Serializer implements methods for encoding/decoding
+// envelopes using protobuf encoding.
+type Serializer struct{}
+
+// func init() {
+// 	wire.SetEnvelopeSerializer(Serializer{})
+// }
```
Comment (matthiasgeihs) [inline wire/protobuf/serializer.go:33]: Hm?

File: wire/protobuf/serializer.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+	if err := env.Sender.UnmarshalBinary(protoEnv.Sender); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling sender address")
+	}
+	env.Recipient = wire.NewAddress()
+	if err := env.Recipient.UnmarshalBinary(protoEnv.Recipient); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling recipient address")
+	}
+
+	switch protoEnv.Msg.(type) {
+	case *Envelope_PingMsg:
```
Comment (matthiasgeihs) [inline wire/protobuf/serializer.go:126]: I realize that this PR is not complete yet as many message types are missing, right?

File: wire/test/msgstest.go [hunk 5]
```diff
@@ -46,8 +46,12 @@ func AuthMsgsTest(t *testing.T, serializerTest func(t *testing.T, msg wire.Msg))
 
 // newRandomstring returns a random ascii string of length between minLen and
 // minLen+maxLenDiff.
-func newRandomString(rng *rand.Rand, minLen, maxLenDiff int) string {
+func newRandomASCIIString(rng *rand.Rand, minLen, maxLenDiff int) string {
```
Comment (matthiasgeihs) [inline wire/test/msgstest.go:49]: You forgot to rename in the function comment.

Could it be a problem if protobuf only supports ASCII?

**Review by matthiasgeihs — COMMENTED** [review]
Realized that my review came to early. Seems like not all messages are covered yet.

File: wire/protobuf/serializer.go (hunk 3 above)
Comment (manoranjith) [reply wire/protobuf/serializer.go:33]: The PR is not yet complete. I will adding a few more commits, that will eventually remove this.

File: wire/protobuf/serializer.go (hunk 4 above)
Comment (manoranjith) [reply wire/protobuf/serializer.go:126]: Yes.

File: wire/test/msgstest.go (hunk 5 above)
Comment (manoranjith) [reply wire/test/msgstest.go:49]: I will rename in the comment.

The reason behind this change is that, protobuf encodes strings using UTF-8 encoding and hence supports all unicode characters. However, not all byte arrays are valid unicode strings (see [here](https://en.wikipedia.org/wiki/UTF-8#Encoding)) and protobuf errors when an invalid unicode string is passed to it.

Generating a random unicode string is a bit tricky, so I have used random ASCII string for tests.

Comment (manoranjith) [reply wire/test/msgstest.go:49]: With perunio encoding, this was not a problem because `io.WriteString`, by default, treats the string as byte array.

File: wire/net/test/connhub_internal_test.go [hunk 6]
```diff
@@ -24,6 +24,7 @@ import (
 	_ "perun.network/go-perun/backend/sim" // backend init
 	wallettest "perun.network/go-perun/wallet/test"
 	"perun.network/go-perun/wire"
+	_ "perun.network/go-perun/wire/protobuf" // wire serialzer init
```
Comment (matthiasgeihs) [inline wire/net/test/connhub_internal_test.go:27]: I don't like that the user now has to do these blank imports every time it uses perun. It's easy to forget. Is there a reminder in case you forget to set the serializer like this? What will happen then?

File: wire/protobuf/generate.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+// limitations under the License.
+
+package protobuf
+
+// This command generates the bindings for serializing wire messages using
+// protocol buffers serialization protocol.
+//
+// It requires protoc and protoc-gen-go to be installed.
+
+//go:generate protoc --go_out=. wire.proto
```
Comment (matthiasgeihs) [inline wire/protobuf/generate.go:22]: Do we document this in the readme?

File: wire/protobuf/serializer.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	case *client.VirtualChannelFundingProposalMsg:
+		protoEnv.Msg, err = fromVirtualChannelFundingProposalMsg(msg)
+	case *client.VirtualChannelSettlementProposalMsg:
+		protoEnv.Msg, err = fromVirtualChannelSettlementProposalMsg(msg)
+	case *client.ChannelUpdateAccMsg:
+		protoEnv.Msg = fromChannelUpdateAccMsg(msg)
+	case *client.ChannelUpdateRejMsg:
+		protoEnv.Msg = fromChannelUpdateRejMsg(msg)
+	case *client.ChannelSyncMsg:
+		protoEnv.Msg, err = fromChannelSyncMsg(msg)
```
Comment (matthiasgeihs) [inline wire/protobuf/serializer.go:72]: Please throw an error in the default case.

File: wire/protobuf/serializer.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+	case *Envelope_VirtualChannelFundingProposalMsg:
+		env.Msg, err = toVirtualChannelFundingProposalMsg(protoMsg)
+	case *Envelope_VirtualChannelSettlementProposalMsg:
+		env.Msg, err = toVirtualChannelSettlementProposalMsg(protoMsg)
+	case *Envelope_ChannelUpdateAccMsg:
+		env.Msg = toChannelUpdateAccMsg(protoMsg)
+	case *Envelope_ChannelUpdateRejMsg:
+		env.Msg = toChannelUpdateRejMsg(protoMsg)
+	case *Envelope_ChannelSyncMsg:
+		env.Msg, err = toChannelSyncMsg(protoMsg)
```
Comment (matthiasgeihs) [inline wire/protobuf/serializer.go:155]: Please throw an error in the default case.

File: wire/test/msgstest.go (hunk 5 above)
Comment (matthiasgeihs) [reply wire/test/msgstest.go:49]: What i meant is that now the behavior of some messages is different for different encodings. Probably not a problem in most cases, but in an edge case somebody might try to encode some information into this message as bytes which will work with the perunio encoding, but not with the protobuf encoding.

**Review by matthiasgeihs — COMMENTED** [review]
A few comments.

File: wire/test/msgstest.go (hunk 5 above)
Comment (manoranjith) [reply wire/test/msgstest.go:49]: Okay, now I get your point. I can see two possible solutions:

1. If these fields can hold arbitrary binary data, then we can change its type in protobuf adapter from string to byte array. Then, this restriction does not apply.
2. If these fields can hold only string data, then perunio adapter could probably validate the string, before encoding.

What do you think @matthias ?

File: wire/net/test/connhub_internal_test.go (hunk 6 above)
Comment (manoranjith) [reply wire/net/test/connhub_internal_test.go:27]: I agree that this could be missed, in that case there will be a panic.

I think this (blank imports) is the way, we expect the user to initialize different backends (wallet, channel etc.,). So extended the same paradigm here.

File: wire/protobuf/generate.go (hunk 7 above)
Comment (matthiasgeihs) [reply wire/protobuf/generate.go:22]: No need to do this in the general README. We also don't do it for the contract bindings.

File: wire/test/msgstest.go (hunk 5 above)
Comment (matthiasgeihs) [reply wire/test/msgstest.go:49]: OK, maybe we can add a comment to the type that this is expected to be a unicode string.

File: wire/controlmsgs_test.go (hunk 2 above)
Comment (manoranjith) [reply wire/controlmsgs_test.go:20]: Yes, it is specific to `perunio`. But, the tests are in wire package because the Encode/Decode functions are implemented in this package.

File: wire/test/msgstest.go (hunk 5 above)
Comment (manoranjith) [reply wire/test/msgstest.go:49]: Added the comment. See these two commits: 57e48716 and f08fb808.

File: wire/protobuf/serializer.go (hunk 9 above)
Comment (manoranjith) [reply wire/protobuf/serializer.go:159]: Add in this commit: d995e181.

File: wire/protobuf/serializer.go (hunk 8 above)
Comment (manoranjith) [reply wire/protobuf/serializer.go:73]: Add in this commit: d995e181.

**Comment by manoranjith** [comment]
@matthiasgeihs  I have addressed all the review comments. I have also rebased the branch and squashed the review commits.

**Review by matthiasgeihs — APPROVED** [review]
👍
